frontend/class_node.py

```diff
@@ -0,0 +1,86 @@
+from z3 import Const, IntSort
+
+
+class ClassNode:
+    """
+    Class representing a node in a tree where each node represents a class, and has
+    references to the base class and subclasses.
+    """
+
+    def __init__(self, name, parent_node, type_sort):
+        self.name = name
+        self.parent_node = parent_node
+        self.children = []
+        self.type_sort = type_sort
+
+    def find(self, name):
+        """
+        Looks a class with the given name in the tree starting at this node. If there is
+        no such class, returns None.
+        """
+        if name == self.name:
+            return self
+        for c in self.children:
+            res = c.find(name)
+            if res:
+                return res
+
+    def __str__(self):
+        return str(self.name)
+
+    def all_children(self):
+        """
+        Returns all transitive child nodes.
+        """
+        result = [self]
+        for c in self.children:
+            result += c.all_children()
+        return result
+
+    def all_super(self):
```

Comment:
Can we call this `all_parents` to make it symmetric with respect to `all_children`? Or any other appropriate combination of symmetric names?